### DISCUSSION TITLE
[FEATURE] How to create a product Html attribute form for users using JSON Schema?

### DISCUSSION THREAD

**Discussion by chatugpt** [amzn/selling-partner-api-models#3940] [General]
How to create a product attribute form for users using JSON Schema? 

Many validations are often performed on user-inputted attributes. 

If you need to generate an HTML form for users to input product information based on a JSON Schema, 

how can this be achieved?

**Comment by chatugpt** [comment]
amazon use json-schema  [JSON Schema Draft 2019-09](https://json-schema.org/draft/2019-09/json-schema-core.html)

https://github.com/json-editor/json-editor support Draft-07  (2020-12 partial support)

**Comment by rugved1991** [comment]
Hi @chatugpt,

You can check out our [Listings sample solution](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings) to learn more about this. The sample solution uses JS to provide a UI based listings solution to play around with the different workflows of the Listings experience.


Best,
Rugved
Solutions Architect, SP-API

**Comment by chatugpt** [comment]
> Hi @chatugpt,
> 
> You can check out our [Listings sample solution](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings) to learn more about this. The sample solution uses JS to provide a UI based listings solution to play around with the different workflows of the Listings experience.
> 
> Best, Rugved Solutions Architect, SP-API

This can only serve as a simple demonstration and is not suitable for formal use. It lacks some property linkages and does not accommodate changes based on conditions.

![image](https://github.com/amzn/selling-partner-api-docs/assets/496419/bb8d6a85-967c-44a6-b50e-1ef4fabf5f5b)

For an average developer, converting such a complex JSON schema into an HTML form is too challenging

**Comment by wszwc** [comment]
> > 你好@chatugpt,
> > 您可以查看我们的[列表示例解决方案](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings)以了解更多相关信息。该示例解决方案使用 JS 提供基于 UI 的列表解决方案，以体验列表体验的不同工作流程。
> > 最佳 Rugved 解决方案架构师，SP-API
> 
> 这只能作为简单的演示，不适合正式使用。它缺乏一些财产联系，并且不适应根据条件的变化。
> 
> ![图像](https://private-user-images.githubusercontent.com/496419/301447613-bb8d6a85-967c-44a6-b50e-1ef4fabf5f5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3MjQwMzYsIm5iZiI6MTcwOTcyMzczNiwicGF0aCI6Ii80OTY0MTkvMzAxNDQ3NjEzLWJiOGQ2YTg1LTk2N2MtNDRhNi1iNTBlLTFlZjRmYWJmNWY1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwNlQxMTE1MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZTgzYWU5MjUzODY5NjIwNzA1ZWJiMWEwNDE2ZmU4YjM4OGRhMmI2ZmE3MWNlNTAwYThlNWY3YjE0MjQ5NTc1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-UhaD4EZ_um4X50dJH-f7RRBh_ig6COGNGJNK23o1zQ)
> 
> 对于普通开发人员来说，将如此复杂的 JSON 模式转换为 HTML 表单太具有挑战性

Hello, I encountered the same problem, do you have a solution? The problem I encountered is as follows
1. How to filter required attributes for users to fill in like Amazon backend
2. How do variant themes correspond to specific attributes? In the returned JSON Schema, there are more than 60 enumerations of variant themes, and the corresponding specific attributes cannot be linked. This leads me to not knowing how to allow users to create multi-variant items
3. How to effectively use the verification of the "anyof" part?

**Reply by Flora-XueXinyi** [reply]
+1

**Comment by zhaoydo** [comment]
Same problem

**Reply by weilinggu** [reply]
Hi @zhaoydo 

Here are some resources available to help speed up Product Type JSON schema implementation:

- [Here](https://youtu.be/fraQT3UhFII?si=K4tCPUOA3gILc7mU) is a deep dive video on product type JSON schema that explains how the product type JSON schema works with a demo submission in part II
- [Here](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings) is a Listings sample application that implemented a dynamic UI based on product type JSON schema.

[json-schema.org](https://json-schema.org/tools?query=&sortBy=name&sortOrder=ascending&groupBy=toolingTypes&licenses=&languages=&drafts=&toolingTypes=&environments=) has tools you can use to render the product type JSON schema and validate your product payload. We recommend you validate your product payload using these tools first before submitting the product payload. You can also consider using VALIDATION PREVIEW with putListingsItem to test a few product payloads before submission.

Hope this helps! Let me know if you have any questions about the information above.

Thanks
Weiling
Amazon Selling Partner Services
